tpe/lib/src/Model.cc

```diff
@@ -75,6 +76,16 @@ Entity &Model::AddLink()
   return *it->second.get();
 }
 
+//////////////////////////////////////////////////
+std::size_t Model::GetLinkCount() const
+{
+  const auto &children = this->GetChildren();
+  return std::count_if(children.begin(), children.end(), [](auto _child)
+  {
+    return std::dynamic_pointer_cast<Link>(_child.second);
```

Comment:
I'm somewhat alarmed that `GetLinkCount` would involve an O(N) iteration through an array and attempting a dynamic cast on each entry. I would strongly urge us to separate the `children` container into separate containers for links and nested models.

The way it is, I think users may get irritated that their very simple for-loop actually has O(N^2) complexity because checking the terminal condition has O(N) complexity on each iteration.

Reply:
Since the `children` container is in the base class `tpelib::Entity`, I opted for adding extra bookkeeping in `tpelib::Model`. https://github.com/ignitionrobotics/ign-physics/pull/229/commits/e17aed3bb5b6ae4e68658cb7ec39dbe4cdd72a16.

Reply:
This adds a little more complexity when it comes to removing models. My approach in #230 is to override the functions `tpelib::Entity::RemoveChildById` and `tpelib::Entity::RemoveChildByName` in `tpelib::Model`. There is some extra work now when removing models, but it's a trad-off I'm willing to accept to keep `GetLinkCount` O(1).